libs/langchain/langchain/tools/github/utils.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from github import GithubException
+from github.Issue import Issue
+
+from langchain.chat_models import ChatOpenAI
+from langchain.chat_models.base import BaseChatModel
+from langchain.prompts.chat import (
+    ChatPromptTemplate,
+    HumanMessagePromptTemplate,
+    SystemMessagePromptTemplate,
+)
+
+if TYPE_CHECKING:
+    from github.Issue import Issue
+    from langchain.chat_models.base import BaseChatModel
+
+
+def generate_branch_name(issue: Issue, llm: BaseChatModel=None):
+    """
+    Helper functions. Use `generate_branch_name()` to generate a meaningful 
+    branch name that the Agent will use to commit it's new code against. 
+    Later, it can use this branch to open a pull request.
+    """
+    if not llm: 
+        llm = ChatOpenAI(temperature=0, model="gpt-4")
+
+    system_template = "You are a helpful assistant that writes clear and concise" \
+                      "GitHub branch names for new pull requests."
+    system_message_prompt = SystemMessagePromptTemplate.from_template(system_template)
+    example_issue = {
+        "title": "Implement an Integral function in C",
+        "body": "This function should take as input a mathematical function " \
+              "and the limits of integration and return the integral value.",
+    }
+
+    prompt = HumanMessagePromptTemplate.from_template(
+        "Given this issue, please return a single string that would be a suitable " \
+        "branch name on which to implement this feature request. Use common software " \
+        "development best practices to name the branch.\n" \
+        "Follow this formatting exactly:" \
+        "Issue: {example_issue}" \
+        "Branch name: `add_integral_in_c`\n\n" \
+        "Issue: {issue}" \
+        "Branch name: `" 
+    )
+
+    # Combine into a Chat conversation
+    chat_prompt = ChatPromptTemplate.from_messages([system_message_prompt, prompt])
+    formatted_messages = chat_prompt.format_messages(
+        issue=str(issue), example_issue=str(example_issue)
+    )
+
+    output = llm(formatted_messages)
+    return _ensure_unique_branch_name(
+        issue.repository, _sanitize_branch_name(output.content)
+    )
+
+
+def _sanitize_branch_name(text):
+    """
+    # Remove non-alphanumeric characters, use underscores.
+    Example:
+        cleaned_text = strip_n_clean_text("Hello, World! This is an example.")
+        print(cleaned_text)  # Output: "Hello_World_This_is_an_example"
+
+    Returns:
+        str: cleaned_text
+    """
+    cleaned_words = [
+        "".join(c for c in word if c.isalnum() or c == "_") for word in text.split()
```

Comment:
don't need `or c == "_"` Condition result is the same without this check.
It is not important. Feel free to ignore :)

Reply:
After a quick test, I think that condition is necessary IMO. It helps when the input text already contains underscores, and we will preserve them for readability.

Without the `or c == "_"` check: 
```python
_sanitize_branch_name("my_branch_name") # mybranchname     BAD because I want underscores
_sanitize_branch_name("my branch name") # my_branch_name   GOOD more readable
```

With the `or c == "_"` check: 
```python
_sanitize_branch_name("my_branch_name") # my_branch_name  GOOD
_sanitize_branch_name("my branch name") # my_branch_name  GOOD
```